File: widgets/Surface3dVtp/Surface3dVtp.py
from mountaintools import client as mt
from vtk.util.numpy_support import vtk_to_numpy
from vtk import vtkXMLPolyDataReader


class Surface3dVtp:
    def __init__(self):
        super().__init__()

    def javascript_state_changed(self, prev_state, state):
        self.set_state(dict(status='running', status_message='Running'))

        vtp_path = state.get('vtp_path', None)
        download_from = state.get('download_from', None)
        scalar_info = state.get('scalar_info', None)
        vector_field_info = state.get('vector_field_info', None)
        arrow_subsample_factor = state.get('arrow_subsample_factor', None)

        if not vtp_path:
            self.set_state(dict(status='error', status_message='No vtp_path'))
            return
        
        if download_from:
            mt.configDownloadFrom(download_from)
        fname = mt.realizeFile(path=vtp_path)

        reader = vtkXMLPolyDataReader()
        reader.SetFileName(fname)
        reader.Update()
        X = reader.GetOutput()
        vertices = vtk_to_numpy(X.GetPoints().GetData()).T
        faces = vtk_to_numpy(X.GetPolys().GetData())

        if scalar_info:
            scalars = vtk_to_numpy(X.GetPointData().GetArray(scalar_info['name']))
            scalars = scalars[:, scalar_info['component']]
        else:
            scalars = None

        if vector_field_info:
            vector_field = vtk_to_numpy(X.GetPointData().GetArray(vector_field_info['name']))
            vector_field = vector_field[:, vector_field_info['components']]
            arrows = [
                dict(
                    start=vertices[:, j] - vector_field[j, :].T / 2,
                    end=vertices[:, j] + vector_field[j, :].T / 2
                )
                for j in range(0, vector_field.shape[0], arrow_subsample_factor)
            ]                
        else:
            arrows = None
            
        self.set_state(dict(
            status='finished',
            vertices=vertices,
            faces=faces,
            scalars=scalars,
            arrows=arrows
        ))

File: widgets/Surface3dVtp/Surface3dVtp.js
import React, { Component } from 'react';
import { PythonInterface } from 'reactopya';
import Surface3d from '../Surface3d/Surface3d';
const config = require('./Surface3dVtp.json');

export default class Surface3dVtp extends Component {
    static title = 'Surface3dVtp'
    static reactopyaConfig = config
    constructor(props) {
        super(props);
        this.state = {
            // javascript state:
            vtp_path: null,
            download_from: null,
            scalar_info: null,
            vector_field_info: null,
            arrow_subsample_factor: null,

            // python state:
            status: '',
            status_message: '',
            vertices: null,
            faces: null,
            scalars: null,
            arrows: null
        }
    }
    componentDidMount() {
        this.pythonInterface = new PythonInterface(this, config);
        this.pythonInterface.setState({
            vtp_path: this.props.vtp_path || '',
            download_from: this.props.download_from,
            scalar_info: this.props.scalar_info,
            vector_field_info: this.props.vector_field_info,
            arrow_subsample_factor: this.props.arrow_subsample_factor
        });
        this.pythonInterface.start();
    }
    componentWillUnmount() {
        this.pythonInterface.stop();
    }
    render() {
        return (
            <RespectStatus {...this.state}>
                <Surface3d
                    {...this.props}
                    vertices={this.state.vertices}
                    faces={this.state.faces}
                    scalars={this.state.scalars}
                    arrows={this.state.arrows}
                />
            </RespectStatus>
        )
    }
}

class RespectStatus extends Component {
    state = {}
    render() {
        switch (this.props.status) {
            case 'running':
                return <div>Running: {this.props.status_message}</div>
            case 'error':
                return <div>Error: {this.props.status_message}</div>
            case 'finished':
                return this.props.children;
            default:
                return <div>Unknown status: {this.props.status}</div>
        }
    }
}
